File: src/FileIO/FileIO.py
"""
/*******************************************************************************
*  Copyright by the contributors to the Dafny Project
*  SPDX-License-Identifier: MIT
*******************************************************************************/
"""

import pathlib
import sys
import traceback

from _dafny import Seq, string_of

assert "DafnyLibraries" == __name__
DafnyLibraries = sys.modules[__name__]

class FileIO:
    @staticmethod
    def INTERNAL_ReadBytesFromFile(path: Seq) -> (bool, Seq, Seq):
        """Attempt to read all bytes from the file at the given path, and return the following values:

        * ``isError``: ``True`` iff an exception was raised during path string conversion or when reading the file
        * ``bytesRead``: the sequence of bytes read from the file, or an empty sequence if ``isError`` is ``True``
        * ``errorMsg``: the error message of the raised exception if ``isError`` is ``True``, or an empty sequence otherwise

        We return these values individually because ``Result`` is not defined in the runtime but instead in library code.
        It is the responsibility of library code to construct an equivalent ``Result`` value.
        """
        try:
            with open(string_of(path), "rb") as file:
                return False, Seq(file.read()), Seq()
        except Exception:
            return True, Seq(), Seq(traceback.format_exc())

    @staticmethod
    def INTERNAL_WriteBytesToFile(path: Seq, bytes_: Seq) -> (bool, Seq):
        """Attempt to write all given bytes to the file at the given path, creating parent directories as necessary,
        and return the following values:

        * ``isError``: ``True`` iff an exception was raised during path string conversion or when writing to the file
        * ``errorMsg``: the error message of the raised exception if ``isError`` is ``True``, or an empty sequence otherwise

        We return these values individually because ``Result`` is not defined in the runtime but instead in library code.
        It is the responsibility of library code to construct an equivalent ``Result`` value.
        """
        try:
            path_str = string_of(path)
            _create_parent_dirs(path_str)
            with open(path_str, "wb") as file:
                file.write(bytes(bytes_))
                return False, Seq()
        except Exception:
            return True, Seq(traceback.format_exc())

    def _create_parent_dirs(path: str) -> str:
        """Create the nonexistent parent directory(-ies) of the given path."""
        pathlib.Path(path).absolute().parent.mkdir(parents=True)

File: src/FileIO/FileIO.js
const fs = require("fs");
const buffer = require("buffer");
var DafnyLibraries = DafnyLibraries || {};
DafnyLibraries.FileIO = (function() {
  let $module = {};

  /**
   * Attempts to read all bytes from the file at the given `path`, and returns an array of the following values:
   *
   *   - `isError`: true iff an error was thrown during path string conversion or when reading the file
   *   - `bytesRead`: the sequence of bytes from the file, or an empty sequence if `isError` is true
   *   - `errorMsg`: the error message of the thrown error if `isError` is true, or an empty sequence otherwise
   *
   * We return these values individually because `Result` is not defined in the runtime but instead in library code.
   * It is the responsibility of library code to construct an equivalent `Result` value.
   */
  $module.INTERNAL_ReadBytesFromFile = function(path) {
    const emptySeq = _dafny.Seq.of();
    try {
      const readOpts = { encoding: null };  // read as buffer, not string
      const buf = fs.readFileSync(path, readOpts);
      const readBytes = _dafny.Seq.from(buf.valueOf(), byte => new BigNumber(byte));
      return [false, readBytes, emptySeq];
    } catch (e) {
      const errorMsg = _dafny.Seq.from(e.toString());
      return [true, emptySeq, errorMsg];
    }
  }

  /**
   * Attempts to write all given `bytes` to the file at the given `path`, and returns an array of the following values:
   *
   *   - `isError`: true iff an error was thrown during path string conversion or when writing to the file
   *   - `errorMsg`: the error message of the thrown error if `isError` is true, or an empty sequence otherwise
   *
   * We return these values individually because `Result` is not defined in the runtime but instead in library code.
   * It is the responsibility of library code to construct an equivalent `Result` value.
   */
  $module.INTERNAL_WriteBytesToFile = function(path, bytes) {
    try {
      const buf = buffer.Buffer.from(bytes);
      fs.writeFileSync(path, buf);  // no need to specify encoding because data is a Buffer
      return [false, _dafny.Seq.of()];
    } catch (e) {
      const errorMsg = _dafny.Seq.from(e.toString());
      return [true, errorMsg];
    }
  }

  return $module;
})();
